### PR TITLE
Add Rotary Positional Embeddings (RoPE) - part 2 of parallel attention blocks

### PR DESCRIPTION
**Pull request by lessw2020**
Summary:
Adds Rotary Positional Embeddings (RoPE)

Test plan:
two unit tests - one for math, one for padding

### CODE REVIEW

File: torchmultimodal/modules/layers/position_embedding.py [hunk 1]
```diff
@@ -169,3 +169,105 @@ def forward(self, t: Tensor) -> Tensor:
...
+            kv-caching and left-padding situations, for which the rotation to be applied might
+            not always be the pre-cached position 0...S. For kv-caching without dynamic batching
+            start_pos is shared for all the batch.
+        """
+        seq_len = q.shape[2]
+        q_ = q.float().reshape(*q.shape[:-1], -1, 2)  # B H L D/2 2
+        k_ = k.float().reshape(*k.shape[:-1], -1, 2)  # B H L D/2 2
+
+        if isinstance(start_pos, int):
+            if start_pos + seq_len > self.max_seq_len_cached:
```
Comment (ebsmothers) [inline torchmultimodal/modules/layers/position_embedding.py:257]: Some comments here about when the frequencies need to be recomputed might be helpful

File: torchmultimodal/modules/layers/position_embedding.py [hunk 2]
```diff
@@ -169,3 +169,105 @@ def forward(self, t: Tensor) -> Tensor:
...
+            "freqs",
+            1.0
+            / (
+                ratio
+                ** (torch.arange(0, dim, 2, device=device)[: (dim // 2)].float() / dim)
+            ),
+        )
+        self.compute_freqs_cis(max_position_embeddings)
+
+    def compute_freqs_cis(
```
Comment (ebsmothers) [inline torchmultimodal/modules/layers/position_embedding.py:206]: Random q: what does cis mean here?

**Review by ebsmothers — COMMENTED** [review]
Looks good! Just a few minor things, mainly around testing and comments

**Review by rohan-varma — COMMENTED** [review]
high level comment, but let's maybe create a modules/layers/embeddings folder in the future as we might have multiple embedding layers.

File: tests/modules/layers/test_position_embedding.py [hunk 3]
```diff
@@ -112,3 +115,38 @@ def test_forward(self, data, emb):
         actual = emb(data)
         expected = torch.Size([3, 5])
         assert_expected(actual.shape, expected)
+
+
+def test_rotary_embeddings_math():
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply tests/modules/layers/test_position_embedding.py:120]: yes, will do.

File: tests/modules/layers/test_position_embedding.py [hunk 4]
```diff
@@ -112,3 +115,38 @@ def test_forward(self, data, emb):
...
+    rotary_embeddings = RotaryPositionalEmbeddings(16, 32)
+
+    qr, kr = rotary_embeddings(q, k, 0)
+    qr2, kr2 = rotary_embeddings(q, k, torch.tensor([0, 1]))
+
+    assert_expected(qr[0], qr2[0])
+    assert_expected(qr[0, :, 1], qr2[1, :, 0])
+
+    assert_expected(kr[0], kr2[0])
+    assert_expected(kr[0, :, 1], kr2[1, :, 0])
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply tests/modules/layers/test_position_embedding.py:152]: yes, that's a good idea.

File: torchmultimodal/modules/layers/position_embedding.py (hunk 2 above)
Comment (lessw2020) [reply torchmultimodal/modules/layers/position_embedding.py:206]: it's short form for rotation transform technically doing e^(alpha*i) = cos(alpha) + i * sin(alpha), or shortened, **c**os + **i** * **s**in = cis.

Comment (lessw2020) [reply torchmultimodal/modules/layers/position_embedding.py:206]: should probably add that in the docstring actually, otherwise it's too cryptic.

File: torchmultimodal/modules/layers/position_embedding.py (hunk 1 above)
Comment (lessw2020) [reply torchmultimodal/modules/layers/position_embedding.py:260]: sounds good - offhand should be changing dtype, changing device, and resetting seq len > max_seq_len.
